unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let (depth, name) = match link.Name {
+        Some(Cow::Borrowed(s)) if s.len() > 2 => (depth, s[2..].to_owned()),
+        Some(Cow::Borrowed(s)) if s.len() == 2 => (depth - 1, String::from("")),
+        None | Some(Cow::Borrowed(_)) => todo!("link cannot be empty"),
+        Some(Cow::Owned(s)) => {
+            if s.len() == 2 {
+                (depth - 1, String::from(""))
+            } else {
+                assert!(s.len() > 2);
+                (depth, s[2..].to_owned())
+            }
+        },
+    };
+    assert!(!name.contains('/'));
+    Ok((cid, name, depth))
+}
+
+impl Walker {
+    pub fn start<'a>(data: &'a [u8], cache: &mut Option<Cache>) -> Result<Walk<'a>, Error> {
+        let flat = FlatUnixFs::try_from(data)?;
+        let metadata = FileMetadata::from(&flat.data);
+
+        match flat.data.Type {
+            UnixFsType::Directory => {
+                let inner = InnerEntry::new_root_dir(metadata);
+
+                let links = flat.links
+                    .into_iter()
+                    .enumerate()
+                    // 2 == number of ancestors this link needs to have on the path, this is after
+                    // some trial and error so not entirely sure why ... ancestors always includes
+                    // the empty root in our case.
+                    .map(|(nth, link)| convert_link(2, nth, link));
+
+                Self::walk_directory(links, inner)
+            },
+            UnixFsType::HAMTShard => {
+                let inner = InnerEntry::new_root_bucket(metadata);
+
+                let links = flat.links
+                    .into_iter()
+                    .enumerate()
+                    // 1 == ancestors should be only the [""]
+                    .map(|(nth, link)| convert_sharded_link(1, nth, link));
+
+                Self::walk_directory(links, inner)
+            },
+            UnixFsType::Raw | UnixFsType::File => {
+                let (bytes, metadata, step) = IdleFileVisit::default()
+                    .start_from_parsed(flat, cache)?;
+
+                if let Some(visit) = step {
+                    Ok(Walk::MultiBlockFile {
+                        metadata,
+                        visit
+                    })
+                } else {
+                    Ok(Walk::SingleBlockFile {
+                        bytes,
+                        metadata
+                    })
+                }
+            },
+            UnixFsType::Metadata => todo!("metadata?"),
+            UnixFsType::Symlink => todo!("?"),
+        }
+    }
+
+    fn walk_directory<'a, I>(mut links: I, current: InnerEntry) -> Result<Walk<'a>, Error>
+        where I: Iterator<Item = Result<(Cid, String, usize), InvalidCidInLink>> + 'a,
+    {
+        if let Some(next) = links.next() {
+            let next = Some(next?);
```

Comment:
perhaps these 2 lines could just be made into `if let Some(second) = links.nth(1)`? that seems a bit more clear than `let next = Some(next?);`